Data/ByteString/Internal/Pure.hs

```diff
@@ -0,0 +1,366 @@
+{-# LANGUAGE MagicHash #-}
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE MultiWayIf #-}
+
+-- | Haskell implementation of C bits
+module Data.ByteString.Internal.Pure
+  ( -- * fpstring.c
+    intersperse
+  , countOcc
+  , countOccBA
+  , reverseBytes
+  , findMaximum
+  , findMinimum
+  , quickSort
+  , elemIndex
+  , isValidUtf8
+  , isValidUtf8BA
+  -- * itoa.c
+  , decodeSignedDec
+  , decodeUnsignedDec
+  , decodeUnsignedDecPadded
+  , decodeUnsignedHex
+  -- * static tables (unaligned!)
+  , lower_hex_table
+  , digit_pairs_table
+  )
+where
+
+import Prelude
+
+import GHC.Exts                 (Ptr(..), ByteArray#, indexWord8Array#, Word8#, Int#, indexWord8OffAddr#)
+import GHC.Types                (Int (..))
+import GHC.Word                 (Word8(..))
+import GHC.Int                  (Int8(..))
+
+import Data.Bits                (Bits(..), shiftR, (.&.))
+import Data.Word
+import Foreign.Ptr              (plusPtr)
+import Foreign.Storable         (Storable(..))
+import Control.Monad            (when)
+
+----------------------------------------------------------------
+-- Haskell version of functions in fpstring.c
+----------------------------------------------------------------
+
+-- | duplicate a string, interspersing the character through the elements of the
+-- duplicated string
+intersperse :: Ptr Word8 -> Ptr Word8 -> Int -> Word8 -> IO ()
+intersperse !dst !src !len !w = case len of
+  0 -> pure ()
+  1 -> do
+    -- copy last char
+    c <- peekByteOff src 0
+    pokeByteOff dst 0 (c :: Word8)
+  _ -> do
+    c <- peekByteOff src 0
+    pokeByteOff dst 0 (c :: Word8)
+    pokeByteOff dst 1 w
+    intersperse (plusPtr dst 2) (plusPtr src 1) (len-1) w
+
+countOccBA :: ByteArray# -> Int -> Word8 -> IO Int
+countOccBA ba len w = return (go 0 0)
+  where
+    go !n !i@(I# i#)
+      | i == len = n
+      | W8# (indexWord8Array# ba i#) == w = go (n+1) (i+1)
+      | otherwise = go n (i+1)
+
+countOcc :: Ptr Word8 -> Int -> Word8 -> IO Int
+countOcc p len w
+  | len == 0  = pure 0
+  | otherwise = count_occ (plusPtr p (len - 1)) w 0 p
+
+count_occ :: Ptr Word8 -> Word8 -> Int -> Ptr Word8 -> IO Int
+count_occ !plast !w !count !p = do
+  c <- peekByteOff p 0
+  let !count' = if c == w then count+1 else count
+  if p == plast
+    then pure count'
+    else count_occ plast w count' (plusPtr p 1)
+
+elemIndex :: ByteArray# -> Word8 -> Int -> IO Int
+elemIndex !ba !w !len = return (go 0)
```

Comment:
I don't think the bangs on `!ba` and `!w` are necessary, just like in `countOccBA` these will be scrutinized for each iteration of `go`